File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
# 내위치 1,1 / 상대방 위치n,m
# 상대방 위치 구하기 -> 행, 렬
# 못찾으면 -1 반환
# 좌표 설정 상하좌우
# 재귀, 큐 써서
# 1,1 q에 넣고 방문true
# for 문 한도에 벗어나면 스킵, 상하좌우 길이만큼 만약 다음곳이 1이면 큐에넣기 count +1 -> 전역변수

from collections import deque
def solution(maps):
    answer = 0
    d = [(1,0), (-1,0),(0,-1),(0,1)] #상하좌우
    n, m = len(maps), len(maps[0]) # 상대방 위치 구하기 -> 행, 렬
    dq = deque()
    start = (0,0,1)
    distance = 0
    dq.append(start)
    imi = [[False] * len(maps[0]) for _ in range(len(maps))]
    
    while dq:
        dy, dx, distance = dq.popleft()
        
        if dy == n-1 and dx == m-1:
            return distance
        
        for row, col in d:
            if 0 <= dy+row < n and 0 <= dx+col < m:
                if imi[dy+row][dx+col] == False and maps[dy+row][dx+col] != 0:
                    imi[dy+row][dx+col] = True
                    dq.append((dy+row,dx+col, distance + 1 ))
                    
    return -1


# UnboundLocalError: local variable 'count' referenced before assignment
# RecursionError: maximum recursion depth exceeded while calling a Python object



# visited된 곳이면 false?
# 

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cpp
#include<vector>
#include<queue>
using namespace std;


int solution(vector<vector<int> > maps)
{
    int answer = 0;
    vector<vector<bool>> visited(maps.size(), vector<bool>(maps[0].size(), false));
    queue<vector<int>> q;
    vector<int> dx = {0,0,-1,1};
    vector<int> dy = {1,-1,0,0};
    
    q.push({0,0,1});
    visited[0][0]= true;
    while(!q.empty()){
        vector<int> f = q.front();
        q.pop();
        
        if(f[0] == maps[0].size()-1 && f[1] == maps.size()-1){
            answer = f[2];
        }
    
        for(int i=0; i<4; i++){
            int nx = f[0] + dx[i];
            int ny = f[1] + dy[i];
            if(0<=nx && nx < maps[0].size() && 0 <= ny && ny < maps.size() && visited[ny][nx] == false && maps[ny][nx] == 1){
              visited[ny][nx]= true;
                q.push({nx,ny,f[2]+1});
            }
        }

    }
        
    if(answer ==0){
        return -1;
    }
    return answer;
}



// 큐선언(x,y,count), visited생성, 동서남북전역변수
// 1. 큐에 초기값을 넣는다 push
// 2. while(큐가 안비어있을때까지)
//  - front값 확인
//  - 큐값 pop
//  - 현재 front x,y 값이 maps 사이즈 -1이라면 count값 answer에 넣기
//  - 현재 방문처리
//  - for 동서남북 , map안에있고 방문안했고 1이라면 push(nx,ny,count+1)
// 3. answer=0이면 -1반환 